[frameit] Ensure both texts use same baseline (issue #10440)

funnel20 opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (do: [x], don't: [x ], [ x], [✔️]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
Output: 
```
Finished in 6 minutes 46 seconds (files took 3.7 seconds to load)
4675 examples, 0 failures
```
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
Output: 
```
973 files inspected, no offenses detected
```
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.
Not applicable

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
This solves issue https://github.com/fastlane/fastlane/issues/10440
<!-- If it fixes an open issue, please link to the issue here. -->
The current implementation applies a `.trim` to both images for Keyword and Title.
Due to typography differences (for example the "L" that needs more vertical space than an "e") the generic baseline between 2 texts is lost. This can be seen in the excellent analysis of @janpio in this comment: https://github.com/fastlane/fastlane/issues/10440#issuecomment-348010988

_Before_
![iphone6-01_remotetab_framed](https://user-images.githubusercontent.com/2139844/33526905-fee7d936-d847-11e7-8fba-85fb829fc25d.png)

This solution requests the calculated trim box and offset for each image **without applying it**.
By comparing the vertical top offsets and heights of both trim boxes, the vertical offset between them can be calculated. The box with the biggest offset will be compensated to equal the same of the other, so their text baseline will be equal. After this the height is made equal, so they share the same bottom offset.
Finally a `.crop` will be applied with the (compensated) trim boxes.

_After_
![iphone6-01_remotetab_framed](https://user-images.githubusercontent.com/2139844/33526915-2c586160-d848-11e7-9a26-a0bd6ca78ad9.png)

<!-- Please describe in detail how you tested your changes. -->
Test results can be found in issue comment: https://github.com/fastlane/fastlane/issues/10440#issuecomment-348760923

Here the difference is clear, while highlighting the trim box in yellow:

Original code with `.trim`:
![iphone6-01_remotetab_framed trim](https://user-images.githubusercontent.com/2139844/33548783-a7a2fe0c-d8e8-11e7-9ec5-85ffe32e9172.png)

My initial commit with `.crop` to align top:
![iphone6-01_remotetab_framed vertical top](https://user-images.githubusercontent.com/2139844/33548811-c8363d82-d8e8-11e7-8e18-bef2c5104a06.png)

My robust solution with `.crop` for both align top and align bottom:
![iphone6-01_remotetab_framed vertical bottom](https://user-images.githubusercontent.com/2139844/33548816-cde0f060-d8e8-11e7-961a-4b936f286d69.png)

It also works when only one of the images needs to be adjusted both on the top **and** bottom side. In this case for "en":
![iphone6-01_remotetab_framed](https://user-images.githubusercontent.com/2139844/33549011-7536f846-d8e9-11e7-8e62-c2ad323157c8.png)

### Description
<!-- Describe your changes in detail -->
The changes can be found in issue comment: https://github.com/fastlane/fastlane/issues/10440#issuecomment-348760717

Please note that:

- This will supersede PR #11096 as I committed that anonymously.
- I've no Ruby skills, so this might not be proper Production code. But the description and the code show the intended behaviour.
- Hence @janpio offered to help, so I added [WIP] to the title of the PR.

funnel20 committed:
Solution for Frameit trim issue #10440

janpio commented:
Remaining test failure is discussed here:
https://github.com/fastlane/fastlane/pull/11096#issuecomment-349593731
https://github.com/fastlane/fastlane/pull/11096#issuecomment-349602194

janpio commented:
> The reason why I ended up with .identify was that I tried to do it with MiniMagick::Tool::Identify.new, but couldn't find a way to let that act on the image variable title_image instead of a file name.

Question is if `MiniMagick::Tool::Identify.new` would also use GraphicsMagick or ImageMagick... I don't really understand how it decides that :/ Only makes sense to switch if it does _not_ use GraphicsMagick which seems to be missing as a dependency.

Interesting enough `imagemagick` also doesn't seem to be defined as an dependency:
https://github.com/fastlane/fastlane/search?utf8=%E2%9C%93&q=imagemagick&type=
Does it maybe just work by accident (as it is maybe installed on macOS by default)?

janpio commented:
I misinterpreted the error message:

` ImageMagick/GraphicsMagick is not installed`

means that the one of the both it tried to use is not installed.

funnel20 commented:
> Question is if MiniMagick::Tool::Identify.new would also use GraphicsMagick or ImageMagick... I don't really understand how it decides that :/ Only makes sense to switch if it does not use GraphicsMagick which seems to be missing as a dependency.

Line 327 (existing code before my edits) uses `title_image = MiniMagick::Image.open(empty_path)` to create the image. That did pass the tests before.

funnel20 commented:
> Interesting enough imagemagick also doesn't seem to be defined as an dependency:
> https://github.com/fastlane/fastlane/search?utf8=%E2%9C%93&q=imagemagick&type=
> Does it maybe just work by accident (as it is maybe installed on macOS by default)?

I have looked in my Terminal history. When I followed instructions from https://github.com/fastlane/fastlane/blob/master/YourFirstPR.md#checking-out-the-fastlane-repo
I needed to install Bundler and later in the local Fastlane repo run: `bundle update`
One of many output lines was:
`Using mini_magick 4.5.1`

janpio commented:
Yes, but mini_magick is just a wrapper around ImageMagick and GraphicsMagick. 
But this is in the docs: https://docs.fastlane.tools/getting-started/ios/screenshots/#dependencies

Still strange why it didn't fail before...

I pulled @milch into this, he is looking into it.

milch committed:
Fix tests

funnel20 commented:
Maybe the test script is altered?

milch commented:
@funnel20 I found the issue and fixed the tests directly. It should work now! We'll let it run for now to see if it does. 

It works locally because we all probably have imagemagick installed on our machines, which the CI doesn't have

funnel20 commented:
@milch Awesome, I see in your commit that the script just fakes some output. Thanks for your super fast assistance.

funnel20 commented:
I'm glad to see that the test passed.
@milch But someone here doesn't have a CLA...

janpio commented:
That is something different and we can ignore - it just means more than 1 person commited and the maintainers have to check manually of it's okay to be merged. As @milch is the second person, it's all good :)

Can you remove the [WIP] from the title to indicate it is good to be reviewed?

milch reviewed (approved):
@funnel20 Great PR, thanks for your help 👍 

Sorry that you were having issues with the CLA in the previous PR. I definitely understand that the whole CLA business is frustrating for a first-time contributor. I hope now that everything is sorted out it won't discourage you from helping out on fastlane other Google-owned open source projects

To satisfy our @googlebot overlord, @funnel20 has agreed to have their code incorporated by submitting this PR and I, the only other commit author in this PR, obviously also agree to have my commit merged

milch commented:
@funnel20 As @janpio said we can ignore the bot in this case. We both signed the CLAs and it is obvious that I agree to having the commit I made merged as well. 

The [WIP] was just because the tests didn't pass, right? The code looks good to me so if you confirm that we'll merge

funnel20 commented:
I have removed the WIP. It was because I have no Ruby skills and did the review together with @janpio in an earlier PR: https://github.com/fastlane/fastlane/pull/11096#pullrequestreview-80704510

funnel20 commented:
About the CLA. My blood pressure was finally a bit down. 
**But what did just happen?**
Now there is a failing CLA check and we can just override/ignore it by commenting that we all agree.
Or worse, @milch writes this on my behalf. Sorry?
This is unbelievable, unacceptable and probably illegal.

To satisfy the googlebot I would strongly recommend to make a valid, transparent and LEGAL option to let this check pass. Legal because I personally approved for God knows what when signing my CLA.
I expect that each submitter comments a similar command as `I signed it!`, for instance` I agree this PR` and the bot does the math if everyone has done so.

The whole idea of checks is that they should pass. During my endeavour of solving this issue, I hit 3 failing checks:

1. CLA
2. Rubocop error max. amount of Class lines
3. circle failing check.

I did really have to stretch my perseverance by counting till 10 a dozen of times.
But now the last check is just ignored. It's still not green. Or should I say RED.

janpio commented:
Read what @googlebot writes in https://github.com/fastlane/fastlane/pull/11124#issuecomment-349645734

Background is that @googlebot is made for PRs with 1 committer, which covers 99% of PRs that come in. Only when someone with commit rights to the fastlane repo (as @milch and me for example) comes along and decide to commit to your PR, @googlebot bails out and posts this message.

It explicitly mentions that "It's up to you to confirm consent of the commit author(s) and merge this pull request when appropriate.". This is what we can now do to merge this PR: 

1. As @milch is a maintainer, I know he has a signed CLA. 
2. We know that you also have one, as @googlebot said so above.

So technically we could have merged your PR even before you had a valid CLA. But the message from @googlebot told us that this was not okay. Now he tells us, that we can use judgement and thinking to decide if we want to ignore the `x` the check provides right now.

No need to get riled up ;)

> The whole idea of checks is that they should pass.

Actually, the idea is that they check something. And depending on how they are configured, the can block a merge - or just inform about their state (and so optionally be ignored).

funnel20 commented:
So many checks, procedures and rules.
And the last one can stay red.
Seems a bad excuses to not have a proper procedure for this 1% use-case.

ohayon commented on frameit/lib/frameit/editor.rb:
```diff
@@ -347,10 +350,64 @@ def build_title_images(max_width, max_height)
-        title_image.trim # remove white space
+
+        # Trimming the image will result in the loss of the common baseline between the text in all images.
+        # title_image.trim # remove white space
```
Should we just remove these lines instead of leaving it commented?

funnel20 replied:
I left it in for the code review as it was the root cause for the issue.
I agree that the commented `.trim` can be removed.
I suggest to merge the comment "Trimming the image will result..." in the comment a few lines below that starts with "Hence retrieve the calculated...", since it's the rational for the crop solution.

janpio replied:
Maybe make the comment "Naively trimming the image with `.trim` will..."?

ohayon replied:
That's cool @funnel20, you can make that whatever you'd like, just wanted to make sure you thought it was necessary.

funnel20 replied:
@janpio 👍

funnel20 replied:
# Natively trimming the image with .trim will result in the loss of the common baseline between the text in all images. Hence retrieve the calculated trim bounding box without actually trimming. See: http://www.imagemagick.org/script/escape.php

ohayon replied:
thats great @funnel20 !

janpio replied:
I actually really meant "Naively", as in "using the most obvious function that comes to mind to do the job of trimming". But yours probably works as well ;)

funnel20 replied:
@janpio I see :-) 
It's a thinker though, as one might suspect a typo (like I did)

ohayon commented on frameit/lib/frameit/editor.rb:
```diff
@@ -347,10 +350,64 @@ def build_title_images(max_width, max_height)
+        trim_values = calculated_trim_box.split(/[\sx+]/).map(&:to_i) # The format is "wxh+X+Y", so use multiple string separators: the whitespace "\s", "x" and "+"
+
+        # Get the minimum top offset of the trim box:
+        if trim_values[3] < top_vertical_trim_offset
```
This seems a bit magical to know that we should index in to `3` and `1` below, is there any explanation we could add here to make this a bit more clear?

funnel20 replied:
What about this extended comment:
```Ruby
# The .identify with format("%@") returns a string with "<width>x<height>+<offset_x>+<offset_y>". Extract these 4 parameters into an integer array, by using multiple string separators: the whitespace "\s", "x" and "+".
```

To also answer your next comment. Sure, we can define these indexes.
Do you have any naming convention for these kind of constants?
Example:
```Ruby
index_trim_box_width = 0
index_trim_box_height = 1
index_trim_box_offset_x = 2
index_trim_box_offset_y = 3
```

ohayon replied:
I think code that is self explanatory is better than comments, I would say that you could do something like: 

```
trim_box_width = trim_values[0]
trim_box_height = trim_values[1]
trim_box_offset_x = trim_values[2]
trim_box_offset_y = trim_values[3]
```

funnel20 replied:
Understood.

funnel20 replied:
Basically only need trim_box_height and trim_box_offset_y, this saves 2 lines already 👊

funnel20 replied:
Hmm, unfortunately I have to define them twice, because the trim box is used in 2 loops:
1. The loop where the image is created. To get the minimum top and maximum bottom offset of all images.
2. The loop where the trim boxes are compensated based on the earlier minimum top and maximum bottom offsets, to maintain the text baseline.

So now I have this with 2 times `# Get the height and vertical offset`:
```Ruby
        # The .identify with format("%@") returns a string with "<width>x<height>+<offset_x>+<offset_y>". Extract these 4 parameters into an integer array, by using multiple string separators: the whitespace "\s", "x" and "+":
        trim_values = calculated_trim_box.split(/[\sx+]/).map(&:to_i)
                            
        # Get the height and vertical offset:
        trim_box_height = trim_values[1]
        trim_box_offset_y = trim_values[3]

        # Get the minimum top offset of the trim box:
        if trim_box_offset_y < top_vertical_trim_offset
          top_vertical_trim_offset = trim_box_offset_y
        end

        # Get the maximum bottom offset of the trimbox, this is the top offset + height:
        if (trim_box_offset_y + trim_box_height) > bottom_vertical_trim_offset
          bottom_vertical_trim_offset = trim_box_offset_y + trim_box_height
        end

        # Store for the crop action:
        trim_boxes[key] = trim_values
      end

      # Crop images based on top_vertical_trim_offset to maintain text baseline:
      words.each do |key|
        # Get matching trim box:
        trim_box = trim_boxes[key]

        # Get the height and vertical offset:
        trim_box_height = trim_box[1]
        trim_box_offset_y = trim_box[3]
```

ohayon replied:
Okay. So we have a couple options. I don't want you to keep having to make changes as you've already put in a great deal of work here. I am cool to merge it as is, and we can go back in and clean it up later.

The other option would be to create some sort of Ruby object that encapsulates these box dimensions and then have an initializer where it could be created from the array of `trim_boxes` that way it could be created and used a lot more efficiently *and* have the extra benefit of being abstracted into another class and therefor maybe help with the 320 line restriction.

What are your thoughts?

funnel20 replied:
Option 2 is definitely preferred, as it improves on 3 subjects:
- readability
- 320 line check
- re-use

It would be great if you can make it, as I lack the Ruby skills. Or show me a similar file in the repo that I can use as example.

My functional thoughts:
1. The initializer uses the raw string from the .identify as input and retrieves it's 4 contents to create the trim box object with 4 populated properties: width, height, offset_x and offset_y
2. The current Class can read and change these object properties 
3. A method that returns the string format for the `.crop` method.

ohayon replied:
Yeah that sounds awesome and we are definitely happy to help with that!

funnel20 replied:
For me just the outline of a class with an init method, a property and 1 method that returns a string is enough. Then I can populate it all myself.

What do you prefer as Class name?
Although currently everything is titled "trim box", maybe "crop box", "crop frame" or "crop canvas" are more appropriate?

ohayon commented on frameit/lib/frameit/editor.rb:
```diff
@@ -347,10 +350,64 @@ def build_title_images(max_width, max_height)
+          # Increase the height of the trim box (2nd parameter) with the difference in vertical top offset:
+          trim_box[1] += trim_box[3] - top_vertical_trim_offset
+          # Change the vertical top offset to match that of the others:
+          trim_box[3] = top_vertical_trim_offset
```
it seems like we use index 1 and 3 quite a bit, could we stick them into their own, more descriptive variables?

funnel20 replied:
See previous comment

funnel20 replied:
Although..... Adding these 4 parameters would trigger the 320 lines Class limit again.
Or is there an overall Class with constants in the project?

ohayon replied:
Lets not worry about the 320 lines for now. Lets make the code the best we can and we'll worry about trimming it down on our end.

funnel20 replied:
Do you mind if I restore the 1 line that you changed earlier to keep within the 320 limit? We both agreed that didn't result in the best code.

ohayon replied:
Thats cool! I plan to go in and make some changes before merging to keep the lines down. 😄

funnel20 replied:
Can I just commit my changes simultaneously?

ohayon replied:
Sure!

ohayon reviewed (commented):
This is a great change for sure. @funnel20 I just left a few comments that are definitely open for discussion. Nothing is too pressing and we totally understand if you just want to get this in quickly.

funnel20 committed:
Updated after review by @ohayon:
- Removed the original .trim command, which was commented
- Improved comments
- Used local variables instead of the array indexes

funnel20 committed:
Merge branch 'frameit-vertical-text-offset3' of https://github.com/funnel20/fastlane into frameit-vertical-text-offset3

ohayon commented:
@funnel20 disregard that @googlebot message! Apologies for the noise with that.

funnel20 commented:
@ohayon No problem, I think it's my mistake as SourceTree committed the earlier commit of @milch again.

funnel20 commented:
The failing circle is expected, as I exceed the 320 lines (it's 325).
I want to fix that with the new trim box class, but first wanted to make this commit as the final solution in 1 class.

ohayon commented:
Sounds good. Let me know when you are in a place where you want us to take over and handle the refactor into that class! 🐰  🚀

funnel20 commented:
As mentioned in the (collapsed) review comment:

> If you can write (or point to) the outline of a class with an init method, a property and 1 method that returns a string is enough. Then I can populate it all myself.
> 
> What do you prefer as Class name?
> Although currently everything is titled "trim box", maybe "crop box", "crop frame" or "crop canvas" are more appropriate?

I'd like to start it myself, as this code generation gives me finally some satisfaction over all administrative hassle

funnel20 commented:
I guess this is a good candidate: https://github.com/fastlane/fastlane/blob/master/frameit/lib/frameit/screenshot.rb

ohayon commented:
Ah got it, sorry I missed that message! 

Yeah, that screenshot class you linked to is a perfect candidate. I was just drafting up something like this:

```
class TrimBox
  attr_accessor :width
  attr_accessor :height
  attr_accessor :x_offset
  attr_accessor :y_offset

  def initialize(idenetify_string)
    @width = ...
    
  end

  def crop
    return "some string"
  end
end
```

funnel20 commented:
Awesome, try that tomorrow locally.
Two Ruby newbie questions:
1. In the screenshot class I sometimes see the @ before the variable, and sometimes not. What's the rule?
2. How to call the init from the Editor class?

ohayon commented:
Those are two great questions: 

1. the `@` indicates that its an instance variable. Also worth noting that when you have an `attr_accessor` defined with the same name, any references to `variable_name` are really just getting out the value from `@variable_name`
2. just calling `new` on a class will call it's initializer (i.e. `TrimBox.new("some string")` calls through to `TrimBox#initialize`

🚀

funnel20 committed:
To solve the failing circle test "Editor Class has too many lines. [325/320]":
- Created a new Class Trimbox:
* It has 4 properties: @width, @height, @offset_x and @offset_y.
* These are initialised in initialize(identify_string), where the MiniMagick identify_string is split into the 4 parameters.
* Fault handling is added to report: empty string and invalid string that doesn't comply to the expected MiniMagick syntax
* Function string_format() will return the 4 parameters in the MiniMagick string format.
- Used the new Class Trimbox in Editor. This saves the converting operations which are done in the new Class. This also saves the steps (in both loops) to extract the paramaters from the trim box array. As a result multiple lines of code can be removed, so editor.rb will pass the circle test for the 320 line limit.
- Added file name of new Class Trimbox to frameit.rb

funnel20 commented:
@ohayon I committed the new Trimbox class and the modified code in the Editor class.
Can you please verify it?

funnel20 commented on frameit/lib/frameit/trim_box.rb:
```diff
@@ -0,0 +1,33 @@
+
+    # identify_string: A string with syntax "<width>x<height>+<offset_x>+<offset_y>". This is returned by MiniMagick when using function .identify with format("%@"). It is also required for the MiniMagick .crop function.
+    def initialize(identify_string)
+      UI.user_error!("Trimbox can not be initialised with an empty 'identify_string'.") unless identify_string.length > 0
```
I copied this logic from another Class.
When testing with faulty input, I see the Error in Terminal, but Frameit just continues.
It does not write a framed screenshot.

Is this intended behaviour?

ohayon replied:
It's possible this error is getting rescued somewhere and continuing execution. I would say leave it as is, this class looks 💯 !

funnel20 commented on frameit/lib/frameit/trim_box.rb:
```diff
@@ -0,0 +1,33 @@
+    # Get the trimbox parameters in the MiniMagick string format
+    def string_format
+      # Convert trim box parameters to string with syntax: "<width>x<height>+<offset_x>+<offset_y>":
+      return "#{@width}x#{@height}+#{@offset_x}+#{@offset_y}"
```
For readability I might guess the following is more clear:
`return "%dx%d+%d+%d" % [@width, @height, @offset_x, @offset_y]`

Do you agree?

ohayon replied:
I think what you have here looks great already! Using `"#{variable}` is definitely the preferred way to do string interpolation. 🚀

ohayon commented:
@funnel20 - this is awesome. Thank you so much for putting in the time and effort and dealing with some of the stress around the process, I think I speak for our whole team that we have learned a lot and will work to make this easier and a better experience for contributors. We always strive to make contributing a pleasant experience for folks. 

We would love to see more from you. Please don't hesitate to reach out to us with more questions or for help implementing new features or fixing bugs. 🚀 🐙

funnel20 commented:
Excellent. 
Thanks to all for your patience with me and bringing me up to speed with Ruby.
